src/Game.hs

```diff
@@ -1,78 +1,65 @@
 module Game (
-    playGame
+    playGame,
+    resetScreen
 ) where
 
-import Board (Board, Dimensions(..), SlideError(..), newBoard, slideNumber, isStartingLayout)
+import Board (Board, Dimensions(..), SlideError(..), newBoard, slideNumber, isStartingLayout, Width, Height)
 import System.IO
 import System.Console.ANSI
 import Data.Char
 import Data.List
+import qualified Text.Read
 
 -- | Main "game loop".
 playGame :: IO ()
-playGame = askAboutDimensions
-
--- First success state for playing a game: choose the width for the board size.
-askAboutDimensions :: IO ()
-askAboutDimensions = do
+playGame = do
     putStrLn "Let's choose a board size."
     width  <- getPositiveNumber "How many spots wide should the board be (Q to abort)? "
-    askAboutHeight width
-
--- Second success state: choose the height.
-askAboutHeight :: Maybe Int -> IO ()
-askAboutHeight Nothing = abortGame
-askAboutHeight width   = do
-    height <- getPositiveNumber "How many spots high should the board be (Q to abort)? "
-    verifySaneBoard width height
-
--- Third success state is to validate width and height.
-verifySaneBoard :: Maybe Int -> Maybe Int -> IO ()
-verifySaneBoard Nothing _ = abortGame
-verifySaneBoard _ Nothing = abortGame
-verifySaneBoard (Just width) (Just height)
-    | width * height >= 4 = constructBoard width height
-    | otherwise           = do
-        putStrLn "These dimensions make for an uninteresting game.  Try again."
-        putStrLn ""
-        askAboutDimensions
+    case width of
+      Nothing -> abortGame
+      Just w -> do
+        height <- getPositiveNumber "How many spots high should the board be (Q to abort)? "
+        case height of
+          Nothing -> abortGame
+          Just h -> do
+            let dimens = Dimensions w h
+            if isSaneBoard dimens
+              then do
+                resetScreen
+                board <- newBoard dimens
+                playBoard board
+              else do
+                putStrLn "These dimensions make for an uninteresting game.  Try again.\n"
+                playGame
```

Comment:
This is a perfectly valid alternative pattern, and it is good in term of code being nearby.  I actually much prefer named cases (see https://www.youtube.com/watch?v=CQyt9Vlkbis) and try to avoid the staircase pattern, so I went with the tail recursive state machine style.